src/ser/mod.rs

```diff
@@ -201,7 +215,62 @@ impl<'a> ser::Serializer for &'a mut Serializer {
 
     fn serialize_str(self, v: &str) -> Result<Self::Ok> {
         self.buf.push(b'"');
-        self.buf.extend_from_slice(v.as_bytes());
+
+        // Do escaping according to "6. MUST represent all strings (including object member names) in
+        // their minimal-length UTF-8 encoding": https://gibson042.github.io/canonicaljson-spec/
+        //
+        // We don't need to escape lone surrogates because they are not valid unicode code points, even
+        // if they can exist in JSON or JavaScript strings. As a result, lone surrogates
+        // cannot exist in a Rust String. If they do, the bug is in the String constructor.
+        // An excellent explanation is available at https://www.youtube.com/watch?v=HhIEDWmQS3w
+
+        let mut buf = [0u8; 4]; // a char is up to 4 bytes long
+        for c in v.chars() {
+            match c {
+                '\\' => {
+                    self.buf.push(b'\\');
+                    self.buf.push(b'\\');
+                }
+                '"' => {
+                    self.buf.push(b'\\');
+                    self.buf.push(b'"');
+                }
+                '\u{0008}' => {
+                    self.buf.push(b'\\');
+                    self.buf.push(b'b');
+                }
+                '\u{0009}' => {
+                    self.buf.push(b'\\');
+                    self.buf.push(b't');
+                }
+                '\u{000A}' => {
+                    self.buf.push(b'\\');
+                    self.buf.push(b'n');
+                }
+                '\u{000C}' => {
+                    self.buf.push(b'\\');
+                    self.buf.push(b'f');
+                }
+                '\u{000D}' => {
+                    self.buf.push(b'\\');
+                    self.buf.push(b'r');
+                }
+                '\u{0000}'..='\u{001F}' => {
```

Comment:
This would be 6 chars, right? And above you define `let mut buf = [0u8; 4]`. That should be `[0u8; 6]`, no?

Reply:
Wait, that is `self.buf`, `buf` isn't even used, right?

Reply:
`self.buf` is the output buffer for the whole JSON document. `buf` is a local thing required by the [char::encode_utf8](https://doc.rust-lang.org/std/primitive.char.html#method.encode_utf8) API. I'll rename the later to make that clearer